bayes_opt/bayesian_optimization.py

```diff
@@ -260,21 +283,146 @@ def maximize(self,
-                util.update_params()
-                x_probe = self.suggest(util)
-                iteration += 1
 
-            self.probe(x_probe, lazy=False)
+        # if user specifies a dataset it takes approximated points from it
+        if self._dataset is not None:
+            exact_x_dict = []
+            while not self._queue.empty or iteration < n_iter:
+                try:
+                    x_probe = next(self._queue)
+                except StopIteration:
+                    util.update_params()
+                    x_probe = self.suggest(util)
+                    iteration += 1
+
+                try:
+                    exact_x_dict.append(dict(zip(self._space.keys, x_probe.T)))
+
+                except AttributeError:
+                    exact_x_dict.append(x_probe)
+
+                approximation = self.get_approximation(self._dataset, x_probe)
+                if self._target_column is not None and approximation is not None:
+
+                    self._space.register(approximation[0], approximation[1])
+
+                elif approximation is not None:
+                    self.probe(approximation, lazy=False)
+                else:
+                    self.probe(x_probe, lazy=False)
+
+                if self._bounds_transformer:
+                    self.set_bounds(
+                        self._bounds_transformer.transform(self._space))
+            self.dispatch(Events.OPTIMIZATION_END)
+            self.save_res_to_csv(True, exact_x=exact_x_dict)
+
+        else:
+            while not self._queue.empty or iteration < n_iter:
+                try:
+                    x_probe = next(self._queue)
+                except StopIteration:
+                    util.update_params()
+                    x_probe = self.suggest(util)
+                    iteration += 1
+
+                self.probe(x_probe, lazy=False)
+
+                if self._bounds_transformer:
+                    self.set_bounds(
+                        self._bounds_transformer.transform(self._space))
+
+            self.dispatch(Events.OPTIMIZATION_END)
+            self.save_res_to_csv(False)
+
+    def get_approximation(self, dataset, x_probe):
+        """
+        Method to get from the dataset passed by the user the nearest point to the x_probe point
+
+        Parameters
+        ----------
+
+        dataset: pandas.DataFrame
+            dataset specified by the user
+
+        x_probe: dict
+            point found by the optimization process
+        """
+
+        try:
+            x_array = numpy.array(list(x_probe.values()))
+
+        except AttributeError:
+            x_array = x_probe
+
+        min_distance = None
+        min_index = None
+        approximations = []
+
+        if self._target_column is None:
+            for row in dataset.itertuples():
+                dataset_tuple = []
+                for j in range(x_array.size):
+                    dataset_tuple.append(row[j + 1])
```

Comment:
Did you check if you can directly copy `row` into `dataset_tuple` instead of copying element-wise into a list?

Reply:
now the row is copied directly, I used again itertuples() method because iterrows() returns Series of data, now the for is not present